File: StrukturDataHeap/heap_program.py
import csv


# =========================================
# MIN HEAP
# =========================================
class MinHeap:
    def __init__(self):
        self.heap = []

    def insert(self, data):
        self.heap.append(data)
        self.heapify_up(len(self.heap) - 1)

    def heapify_up(self, index):
        while index > 0:
            parent = (index - 1) // 2

            if self.heap[index][0] < self.heap[parent][0]:
                self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
                index = parent
            else:
                break

    def heapify_down(self, index):
        size = len(self.heap)

        while True:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2

            if left < size and self.heap[left][0] < self.heap[smallest][0]:
                smallest = left

            if right < size and self.heap[right][0] < self.heap[smallest][0]:
                smallest = right

            if smallest != index:
                self.heap[index], self.heap[smallest] = self.heap[smallest], self.heap[index]
                index = smallest
            else:
                break

    def delete_root(self):
        if not self.heap:
            return None

        root = self.heap[0]
        last = self.heap.pop()

        if self.heap:
            self.heap[0] = last
            self.heapify_down(0)

        return root

    def display_ascending(self):
        if not self.heap:
            print("\nMin Heap kosong!")
            return

        data = sorted(self.heap, key=lambda x: x[0])

        print("\n========== DATA ASCENDING ==========")
        for item in data:
            print(f"ID: {item[0]} | Nama: {item[1]}")
        print("====================================")


# =========================================
# MAX HEAP
# =========================================
class MaxHeap:
    def __init__(self):
        self.heap = []

    def insert(self, data):
        self.heap.append(data)
        self.heapify_up(len(self.heap) - 1)

    def heapify_up(self, index):
        while index > 0:
            parent = (index - 1) // 2

            if self.heap[index][0] > self.heap[parent][0]:
                self.heap[index], self.heap[parent] = self.heap[parent], self.heap[index]
                index = parent
            else:
                break

    def heapify_down(self, index):
        size = len(self.heap)

        while True:
            largest = index
            left = 2 * index + 1
            right = 2 * index + 2

            if left < size and self.heap[left][0] > self.heap[largest][0]:
                largest = left

            if right < size and self.heap[right][0] > self.heap[largest][0]:
                largest = right

            if largest != index:
                self.heap[index], self.heap[largest] = self.heap[largest], self.heap[index]
                index = largest
            else:
                break

    def delete_root(self):
        if not self.heap:
            return None

        root = self.heap[0]
        last = self.heap.pop()

        if self.heap:
            self.heap[0] = last
            self.heapify_down(0)

        return root

    def display_descending(self):
        if not self.heap:
            print("\nMax Heap kosong!")
            return

        data = sorted(self.heap, key=lambda x: x[0], reverse=True)

        print("\n========== DATA DESCENDING ==========")
        for item in data:
            print(f"ID: {item[0]} | Nama: {item[1]}")
        print("=====================================")


# =========================================
# LOAD CSV
# =========================================
def load_csv(filename, min_heap, max_heap):
    try:
        with open(filename, mode='r', encoding='utf-8') as file:
            reader = csv.DictReader(file)

            jumlah = 0

            for row in reader:
                id_barang = int(row["ID"])
                nama_barang = row["Nama"]

                data = (id_barang, nama_barang)

                min_heap.insert(data)
                max_heap.insert(data)

                jumlah += 1

            print(f"\n{jumlah} data berhasil dimasukkan ke Min Heap & Max Heap!")

    except FileNotFoundError:
        print("\nFile tidak ditemukan!")

    except KeyError:
        print("\nFormat CSV salah!")
        print("Header harus: ID,Nama")

    except Exception as e:
        print("\nTerjadi error:", e)


# =========================================
# MAIN PROGRAM
# =========================================
min_heap = MinHeap()
max_heap = MaxHeap()

while True:

    print("\n====================================")
    print("       PROGRAM MIN & MAX HEAP")
    print("====================================")
    print("1. Tambah Data")
    print("2. Tampilkan Ascending (Min Heap)")
    print("3. Tampilkan Descending (Max Heap)")
    print("4. Hapus Root Min Heap")
    print("5. Hapus Root Max Heap")
    print("6. Load Data dari CSV")
    print("0. Exit")
    print("====================================")

    pilih = input("Pilih menu: ")

    # =====================================
    # TAMBAH DATA
    # =====================================
    if pilih == "1":

        try:
            id_barang = int(input("Masukkan ID: "))
            nama_barang = input("Masukkan Nama: ")

            data = (id_barang, nama_barang)

            min_heap.insert(data)
            max_heap.insert(data)

            print("\nData berhasil ditambahkan!")

        except:
            print("\nInput tidak valid!")

    # =====================================
    # TAMPIL ASCENDING
    # =====================================
    elif pilih == "2":
        min_heap.display_ascending()

    # =====================================
    # TAMPIL DESCENDING
    # =====================================
    elif pilih == "3":
        max_heap.display_descending()

    # =====================================
    # HAPUS MIN HEAP
    # =====================================
    elif pilih == "4":

        deleted = min_heap.delete_root()

        if deleted:
            print("\nData terhapus dari Min Heap:")
            print(f"ID: {deleted[0]} | Nama: {deleted[1]}")
        else:
            print("\nMin Heap kosong!")

    # =====================================
    # HAPUS MAX HEAP
    # =====================================
    elif pilih == "5":

        deleted = max_heap.delete_root()

        if deleted:
            print("\nData terhapus dari Max Heap:")
            print(f"ID: {deleted[0]} | Nama: {deleted[1]}")
        else:
            print("\nMax Heap kosong!")

    # =====================================
    # LOAD CSV
    # =====================================
    elif pilih == "6":

        filename = input("Masukkan nama file CSV: ")

        load_csv(filename, min_heap, max_heap)

    # =====================================
    # EXIT
    # =====================================
    elif pilih == "0":

        print("\nProgram selesai.")
        break

    else:
        print("\nMenu tidak valid!")

File: StrukturDataHeap/HeapProgram.java
import java.io.*;
import java.util.*;

// ======================================
// CLASS DATA
// ======================================
class Data {
    int id;
    String nama;

    public Data(int id, String nama) {
        this.id = id;
        this.nama = nama;
    }

    @Override
    public String toString() {
        return "ID: " + id + ", Nama: " + nama;
    }
}

// ======================================
// MAIN CLASS
// ======================================
public class HeapProgram {

    // ======================================
    // MIN HEAP & MAX HEAP
    // ======================================

    static PriorityQueue<Data> minHeap =
            new PriorityQueue<>(Comparator.comparingInt(d -> d.id));

    static PriorityQueue<Data> maxHeap =
            new PriorityQueue<>((a, b) -> b.id - a.id);

    // ======================================
    // LOAD CSV
    // ======================================
    static void loadCSV(String filename) {

        try {

            BufferedReader br = new BufferedReader(new FileReader(filename));

            String line;

            // skip header
            br.readLine();

            int jumlah = 0;

            while ((line = br.readLine()) != null) {

                String[] data = line.split(",");

                int id = Integer.parseInt(data[0].trim());
                String nama = data[1].trim();

                Data barang = new Data(id, nama);

                minHeap.add(barang);
                maxHeap.add(barang);

                jumlah++;
            }

            br.close();

            System.out.println("\n" + jumlah +
                    " data berhasil dimasukkan ke Min Heap & Max Heap!");

        } catch (FileNotFoundException e) {

            System.out.println("\nFile tidak ditemukan!");

        } catch (Exception e) {

            System.out.println("\nTerjadi error: " + e.getMessage());
        }
    }

    // ======================================
    // INSERT DATA
    // ======================================
    static void insertData(Scanner input) {

        try {

            System.out.print("Masukkan ID   : ");
            int id = Integer.parseInt(input.nextLine());

            System.out.print("Masukkan Nama : ");
            String nama = input.nextLine();

            Data barang = new Data(id, nama);

            minHeap.add(barang);
            maxHeap.add(barang);

            System.out.println("\nData berhasil ditambahkan!");

        } catch (Exception e) {

            System.out.println("\nInput tidak valid!");
        }
    }

    // ======================================
    // SHOW ASCENDING
    // ======================================
    static void showAscending() {

        if (minHeap.isEmpty()) {
            System.out.println("\nMin Heap kosong!");
            return;
        }

        PriorityQueue<Data> temp =
                new PriorityQueue<>(minHeap);

        System.out.println("\n===== DATA ASCENDING (MIN HEAP) =====");

        while (!temp.isEmpty()) {

            System.out.println(temp.poll());
        }
    }

    // ======================================
    // SHOW DESCENDING
    // ======================================
    static void showDescending() {

        if (maxHeap.isEmpty()) {
            System.out.println("\nMax Heap kosong!");
            return;
        }

        PriorityQueue<Data> temp =
                new PriorityQueue<>(maxHeap);

        System.out.println("\n===== DATA DESCENDING (MAX HEAP) =====");

        while (!temp.isEmpty()) {

            System.out.println(temp.poll());
        }
    }

    // ======================================
    // DELETE MIN HEAP
    // ======================================
    static void deleteMinHeap() {

        if (minHeap.isEmpty()) {

            System.out.println("\nMin Heap kosong!");
            return;
        }

        Data deleted = minHeap.poll();

        System.out.println("\nData berhasil dihapus dari Min Heap:");
        System.out.println(deleted);
    }

    // ======================================
    // DELETE MAX HEAP
    // ======================================
    static void deleteMaxHeap() {

        if (maxHeap.isEmpty()) {

            System.out.println("\nMax Heap kosong!");
            return;
        }

        Data deleted = maxHeap.poll();

        System.out.println("\nData berhasil dihapus dari Max Heap:");
        System.out.println(deleted);
    }

    // ======================================
    // SHOW HEAP
    // ======================================
    static void showHeap() {

        System.out.println("\n===== MIN HEAP =====");
        System.out.println(minHeap);

        System.out.println("\n===== MAX HEAP =====");
        System.out.println(maxHeap);
    }

    // ======================================
    // MAIN PROGRAM
    // ======================================
    public static void main(String[] args) {

        Scanner input = new Scanner(System.in);

        while (true) {

            System.out.println("\n====================================");
            System.out.println("      PROGRAM MIN & MAX HEAP");
            System.out.println("====================================");
            System.out.println("1. Tambah Data");
            System.out.println("2. Tampilkan Ascending (Min Heap)");
            System.out.println("3. Tampilkan Descending (Max Heap)");
            System.out.println("4. Hapus Root Min Heap");
            System.out.println("5. Hapus Root Max Heap");
            System.out.println("6. Load Data dari CSV");
            System.out.println("7. Show Heap");
            System.out.println("0. Exit");
            System.out.println("====================================");

            System.out.print("Pilih menu: ");
            String pilih = input.nextLine();

            // ==================================
            // INSERT
            // ==================================
            if (pilih.equals("1")) {

                insertData(input);
            }

            // ==================================
            // SHOW ASCENDING
            // ==================================
            else if (pilih.equals("2")) {

                showAscending();
            }

            // ==================================
            // SHOW DESCENDING
            // ==================================
            else if (pilih.equals("3")) {

                showDescending();
            }

            // ==================================
            // DELETE MIN
            // ==================================
            else if (pilih.equals("4")) {

                deleteMinHeap();
            }

            // ==================================
            // DELETE MAX
            // ==================================
            else if (pilih.equals("5")) {

                deleteMaxHeap();
            }

            // ==================================
            // LOAD CSV
            // ==================================
            else if (pilih.equals("6")) {

                System.out.print("Masukkan nama file CSV: ");
                String filename = input.nextLine();

                loadCSV(filename);
            }

            // ==================================
            // SHOW HEAP
            // ==================================
            else if (pilih.equals("7")) {

                showHeap();
            }

            // ==================================
            // EXIT
            // ==================================
            else if (pilih.equals("0")) {

                System.out.println("\nProgram selesai.");
                break;
            }

            // ==================================
            // INVALID
            // ==================================
            else {

                System.out.println("\nMenu tidak valid!");
            }
        }

        input.close();
    }
}